lib/kinesis.test.js

```diff
@@ -1,139 +1,212 @@
+    expect(calls.length).toEqual(2)
+    expect(calls[0].Records.length).toEqual(2) // overall=failed, segment=success
+    expect(calls[1].Records.length).toEqual(1) // overall=success
+    expect(await isLocked(overall)).toEqual(true)
+    expect(await isLocked(segment)).toEqual(true)
+  })
+
+  it('unlocks kinesis complete failures', async () => {
+    jest.spyOn(kinesis, '_putRecords').mockImplementation(async () => {
+      throw new Error('Complete failure!')
     })
 
-    it('locks the first digest to be downloaded each day', async () => {
-      const data = {le: '1234', digest: '5678', time: 9, bytes: 10, seconds: 12, percent: 0.4}
-      expect(await kinesis.putImpressionLock(redis, {...data, segment: 0})).toEqual(true)
-      expect(await kinesis.putImpressionLock(redis, {...data, digest: '5679'})).toEqual(false)
-      expect(await kinesis.putImpressionLock(redis, {...data, digest: '5679', segment: 3})).toEqual(false)
-      expect(await kinesis.putImpressionLock(redis, {...data, digest: '5679', time: 86400001})).toEqual(true)
-      expect(await kinesis.putImpressionLock(redis, {...data})).toEqual(true)
-
-      expect(allDatas.length).toEqual(5)
-      expect(JSON.parse(allDatas[0])).toMatchObject({digest: '5678', segment: 0})
-      expect(JSON.parse(allDatas[0]).isDuplicate).toBeUndefined()
-      expect(JSON.parse(allDatas[1])).toMatchObject({digest: '5679', isDuplicate: true, cause: 'digestCache'})
-      expect(JSON.parse(allDatas[2])).toMatchObject({digest: '5679', isDuplicate: true, cause: 'digestCache', segment: 3})
-      expect(JSON.parse(allDatas[3])).toMatchObject({digest: '5679', timestamp: 86400001})
-      expect(JSON.parse(allDatas[3]).isDuplicate).toBeUndefined()
-      expect(JSON.parse(allDatas[4])).toMatchObject({digest: '5678', timestamp: 9})
+    jest.spyOn(log, 'warn').mockImplementation(() => null)
+    expect(await kinesis.putWithLock(redis, [overall])).toMatchObject({failed: 1})
+    expect(await isLocked(overall)).toEqual(false)
+
+    expect(log.warn).toHaveBeenCalledTimes(1)
+    expect(log.warn.mock.calls[0][0].message).toMatch('putRecords failure')
+    expect(log.warn.mock.calls[0][1].Records.length).toEqual(1)
+  })
+
+  it('locks to the first digest downloaded', async () => {
+    const calls = mockPutRecords(true, true)
+    expect(await kinesis.putWithLock(redis, [overall])).toMatchObject({overall: 1})
+    expect(await kinesis.putWithLock(redis, [{...overall, digest: '5679'}])).toMatchObject({overallDups: 1})
+    expect(calls.length).toEqual(2)
+    expect(calls[0].Records.length).toEqual(1)
+    expect(JSON.parse(calls[0].Records[0].Data).isDuplicate).toBeUndefined()
+    expect(calls[1].Records.length).toEqual(1)
+    expect(JSON.parse(calls[1].Records[0].Data).isDuplicate).toEqual(true)
+    expect(JSON.parse(calls[1].Records[0].Data).cause).toEqual('digestCache')
+  })
+
+  it('unfortunately does not unlock the digest for a failed putRecord', async () => {
```

Comment:
but this is why we unlock failures after, yes? I guess that means there is some chance these could be left as locked, if the subsequent unlock doesn't occur or fails?

Reply:
Well, there are actually 2 sorts of locks.  The regular `<listener>.<episode>.<utc-day>` lock.  And then this "digest lock" ... which locks us to only counting a single digest for a listener + utc_day.  This was sort of a temporary measure, as our dovetail recent-redirect cache doesn't yet lock the listener to a digest for the entire day.

That said ... not rolling these "digest locks" back is probably fine.  When we retry the failed records, the digest will be locked to the same impressions/downloads we're retrying.